sonar-core/src/main/java/org/sonar/core/component/Module.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.core.component;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.sonar.core.platform.ComponentContainer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public abstract class Module {
+  private ComponentContainer container;
+
+  public Module configure(ComponentContainer container) {
+    this.container = checkNotNull(container);
+
+    configureModule();
+
+    return this;
+  }
+
+  protected abstract void configureModule();
+
+  protected void add(@Nullable Object object, boolean singleton) {
+    if (object != null) {
+      container.addComponent(object, singleton);
+    }
+  }
+
+  protected <T> T getComponentByType(Class<T> tClass) {
+    return container.getComponentByType(tClass);
+  }
+
+  protected void add(Object... objects) {
```

Comment:
Sounds misleading to have both add(Object, boolean) and add(Object...) that implicitly sets the boolean to true. 
Moreover don't you think that these protected methods aim to duplicate ComponentContainer ? An alternative could be: ```protected abstract void configureModule(ComponentContainer container)```